### ISSUE TITLE
How can i use embedding as input for llama model?

### ISSUE THREAD

**Issue by littletomatodonkey** [NVIDIA/TensorRT-LLM#870]
Hi, how can i use embedding as input for llama model (could be better if just on python interface)? I've seen the  following link, but it seems hard for one who wants to define their own inputs and outputs based on trt-llm, does anyone has better idea? thanks!

- https://github.com/NVIDIA/TensorRT-LLM/issues/310
- https://github.com/NVIDIA/TensorRT-LLM/issues/534

**Comment by byshiue** [comment]
If you mean that you want to replace the `input_ids` by `inputs_embedding` and skip the embedding lookup, TRT-LLM does not support such feature yet.

**Comment by littletomatodonkey** [comment]
Thanks for your reply, i've seen the blip2 dir, but it can not satisify my request since i hope the input is just embedding and output is just hidden states. The inputs and outputs need to be fixed for my scene (mainly in `tensorrt_llm/models/llama` and `tensort_llm/runtime`). I'll try my best to solve that. And maybe you can consider this scene (such as for feature extraction or combine with other frameworks)~

**Comment by littletomatodonkey** [comment]
Hi @byshiue I wrap the trt-llm model interface and it supports embedding input and hidden_states output now, the results is normal for batch size=1(compared with huggingface results).
* However, when i set batch size > 1(such as 4), (the input embeddings are all same for each batch sample) the results is abnormal for the odd batch index, which is as follows. batch[1/3/...] is nan, and batch[0/2/...]'s results are right. Can you help to take a look? Thanks!

```
ipdb> outputs["hidden_states"].reshape(4,-1)
tensor([[-0.0812, -0.1250,  0.3232,  ...,  2.1660,  0.5547,  0.1886],
        [    nan,     nan,     nan,  ...,     nan,     nan,     nan],
        [-0.0812, -0.1250,  0.3232,  ...,  2.1660,  0.5547,  0.1886],
        [    nan,     nan,     nan,  ...,     nan,     nan,     nan]],
       device='cuda:0', dtype=torch.float16)
ipdb> input_embed.reshape(4,-1)
tensor([[-0.0038, -0.0039, -0.0067,  ...,  0.0249,  0.0057, -0.0325],
        [-0.0038, -0.0039, -0.0067,  ...,  0.0249,  0.0057, -0.0325],
        [-0.0038, -0.0039, -0.0067,  ...,  0.0249,  0.0057, -0.0325],
        [-0.0038, -0.0039, -0.0067,  ...,  0.0249,  0.0057, -0.0325]],
       device='cuda:0', dtype=torch.float16)
```


I rewrite the `LLaMAForCausalLM` forward, which is follows.

```py
    def forward(self,
                input_ids,
                position_ids=None,
                use_cache=False,
                last_token_ids=None,
                attention_mask=None,
                kv_cache_params=None,
                attention_params=None,
                hidden_states=None,
                all_reduce_workspace=None,
                prompt_embedding_table: Optional[Tensor] = None,
                prompt_tasks: Optional[Tensor] = None,
                prompt_vocab_size: Optional[Tensor] = None,
                lora_params=None,
                input_embed=None): # added by gry
        hidden_states = super().forward(input_ids, position_ids, use_cache,
                                        attention_mask, kv_cache_params,
                                        attention_params, hidden_states,
                                        all_reduce_workspace,
                                        prompt_embedding_table, prompt_tasks,
                                        prompt_vocab_size, lora_params,
                                        input_embed=input_embed) # added by gry

        if use_cache:
            hidden_states, presents = hidden_states
        
        self.register_network_output(f"hidden_states", hidden_states)
        return hidden_states # lm head is not needed so we just return here.
```

**Comment by byshiue** [comment]
You could mark some intermediate results as output to check their values, and locate where does the issue happen.

**Comment by littletomatodonkey** [comment]
Thanks for your reply, i found that for the 1st layers' gpt_attention output, the output is abnormal, which is variable `context` as follows.  https://github.com/NVIDIA/TensorRT-LLM/blob/c89653021e66ca78c55f02b366f404455bc12e8d/tensorrt_llm/layers/attention.py#L732

```
ipdb> ctx_tensors["layers.0.attention.attn0_context_before_dense"].to_torch().reshape(4,-1)
tensor([[ 1.6317e-03, -1.6332e-04,  3.2711e-03,  ..., -4.0829e-05,
          4.7722e-03, -2.1515e-03],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  ...,  0.0000e+00,
          0.0000e+00,  0.0000e+00],
        [ 1.6317e-03, -1.6332e-04,  3.2711e-03,  ..., -4.0829e-05,
          4.7722e-03, -2.1515e-03],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  ...,  0.0000e+00,
          0.0000e+00,  0.0000e+00]], device='cuda:0', dtype=torch.float16)
```

The norm0 and qkv before `gpt_attention` func  is normal: https://github.com/NVIDIA/TensorRT-LLM/blob/c89653021e66ca78c55f02b366f404455bc12e8d/tensorrt_llm/models/llama/model.py#L151

which is as follows. 

```
ipdb> ctx_tensors["layers.0.norm0"].to_torch().reshape(4,-1)
tensor([[-0.0003, -0.0015,  0.0030,  ..., -0.0035, -0.0010, -0.0010],
        [-0.0003, -0.0015,  0.0030,  ..., -0.0035, -0.0010, -0.0010],
        [-0.0003, -0.0015,  0.0030,  ..., -0.0035, -0.0010, -0.0010],
        [-0.0003, -0.0015,  0.0030,  ..., -0.0035, -0.0010, -0.0010]],
       device='cuda:0', dtype=torch.float16)
ipdb> ctx_tensors["layers.0.attention.qkv_0"].to_torch().reshape(4,-1)
tensor([[ 5.2452e-04,  1.9165e-02,  3.6499e-02,  ..., -4.0829e-05,
          4.7722e-03, -2.1496e-03],
        [ 5.2452e-04,  1.9165e-02,  3.6499e-02,  ..., -4.0829e-05,
          4.7722e-03, -2.1496e-03],
        [ 5.2452e-04,  1.9165e-02,  3.6499e-02,  ..., -4.0829e-05,
          4.7722e-03, -2.1496e-03],
        [ 5.2452e-04,  1.9165e-02,  3.6499e-02,  ..., -4.0829e-05,
          4.7722e-03, -2.1496e-03]], device='cuda:0', dtype=torch.float16)
```


Do you have any idea?

**Comment by littletomatodonkey** [comment]
And i found that abs-sum of attn output for the odd batch index sample is zero.

```
ipdb> ctx_tensors["layers.0.attn"].to_torch().reshape(4,-1)[1].abs().sum()
tensor(0., device='cuda:0', dtype=torch.float16)
```

**Comment by byshiue** [comment]
Then, it might be because some inputs of gpt_attention are wrong. You could check them.

**Comment by littletomatodonkey** [comment]
i checked the gpt attention inputs, it seems that no errors. i prepared input embedding as follows (remove input padding is set as False). is it right?

```py
        if use_embed_as_input:
            input_embed = Tensor(
                name='input_embed',
                dtype=dtype,
                shape=[-1, -1, head_size * num_heads],
                dim_range=OrderedDict([
                    ('batch_size_beam_width', bb_range),
                    ('input_len', inlen_range),
                    ('hidden_size',
                        [head_size * num_heads, head_size *
                        num_heads] if enable_two_optimization_profiles else
                        [head_size * num_heads]),
                ])
            )
```

**Comment by littletomatodonkey** [comment]
@byshiue i was wondering that whether tensorrt-llm has some inner tensor slice and computation logic which i did not notice. do you have any idea?

**Comment by littletomatodonkey** [comment]
@byshiue and if i use input ids with input for two same sample (batch sample), the result is normal.

```
ipdb> ctx_tensors["present_key_value_0"].to_torch().reshape(2,-1)
tensor([[0.0068, 0.2095, 0.0292,  ..., 0.0000, 0.0000, 0.0000],
        [0.0068, 0.2095, 0.0292,  ..., 0.0000, 0.0000, 0.0000]],
       device='cuda:0', dtype=torch.float16)
ipdb> ctx_tensors["input_ids"].to_torch().reshape(2,-1)
tensor([[    1, 29871, 33145, 44057, 32172],
        [    1, 29871, 33145, 44057, 32172]], device='cuda:0',
       dtype=torch.int32)
```

**Comment by littletomatodonkey** [comment]
@byshiue I hack the forward process, which takes two inputs (input ids and embedding), and set embedding for the input_ids branch as 0 (multiplied by 0). The results with different batch index are normal now.

```
ipdb> ctx_tensors["embd"].to_torch().reshape(4,-1) # the input ids' branch embedding
dtensor([[0., 0., -0.,  ..., -0., 0., 0.],
        [0., 0., -0.,  ..., -0., 0., 0.],
        [0., 0., -0.,  ..., -0., 0., 0.],
        [0., 0., -0.,  ..., -0., 0., 0.]], device='cuda:0',
       dtype=torch.float16)
ipdb> outputs["hidden_states"].reshape(4,-1)
tensor([[-0.2974,  0.0779,  2.1230,  ...,  0.7036, -1.8740, -0.2573],
        [-0.2974,  0.0779,  2.1230,  ...,  0.7036, -1.8740, -0.2573],
        [-0.2974,  0.0779,  2.1230,  ...,  0.7036, -1.8740, -0.2573],
        [-0.2974,  0.0779,  2.1230,  ...,  0.7036, -1.8740, -0.2573]],
```

The forward code is as follows.


```py
    def forward(self,
                input_ids,
                position_ids=None,
                use_cache=False,
                attention_mask=None,
                kv_cache_params=None,
                attention_params=None,
                hidden_states=None,
                all_reduce_workspace=None,
                prompt_embedding_table: Optional[Tensor] = None,
                prompt_tasks: Optional[Tensor] = None,
                prompt_vocab_size: Optional[Tensor] = None,
                lora_params=None,
                input_embed=None):

        kv_cache_params.fill_none_tensor_list(len(self.layers))

        if use_cache:
            presents = []

        ptuning_args = []
        if self.use_prompt_tuning:
            ptuning_args = [
                prompt_embedding_table, prompt_tasks, prompt_vocab_size
            ]
        
            hidden_states = self.vocab_embedding(input_ids, *ptuning_args,
                                                all_reduce_workspace)
        else:
            hidden_states = recv(hidden_states, self.mapping.prev_pp_rank())
        hidden_states *= 0.0 # hack here @byshiue
        self.register_network_output(f"embd", hidden_states)

        if self.use_embed_as_input: # it is set as True here
            hidden_states = input_embed + hidden_states
        # other code for forward process.
```

**Comment by OswaldoBornemann** [comment]
Thanks for your great work. So I have the same situation like yours @littletomatodonkey . But I only need to change the input type into `input_embeds`. So can I achieve this? Thank you. 

By the way, I think I would like to use `ModelRunnerCpp` instead.
